doc/guides/keycloak_protection.md

```diff
@@ -1,68 +1,118 @@
+* Fields
+* Queries
+* Mutations
+* Subscriptions
 
-Define your resolvers as you normally would. However, you can now use the properties and methods in `context.auth` object which is of type `AuthContextProvider` defined in `apollo-voyager-server`.
-```javascript
-const resolvers = {
-  Query: {
-    hello: (obj, args, context, info) => {
+The `@hasRole` usage is as follows:
 
-      // log some of the auth related info added to the context
-      console.log(context.auth.isAuthenticated())
+* `@hasRole(role: String)`
+  * Example - `@hasRole(role: "admin"])`
+  * If the authenticated user has the role `admin` they will be authorized.
+* `@hasRole(role: [String])`
+  * Example - `@hasRole(role: ["admin", "editor"])`
+  * If the authenticated user has **at least one of the roles** in the list, they will be authorized.
 
-      const name = context.auth.accessToken.content.name || 'world'
-      return `Hello ${name} from ${context.serverName}`
-    }
-  }
-}
-```
+**The default behaviour is to check client roles.** For example, `@hasRole(role: "admin"])` will check that user has a client role called `admin`. `@hasRole(role: "realm:admin"])` will check if that user has a realm role called `admin` 
 
-Read the Keycloak config and pass it to `KeycloakSecurityService`.
-```javascript
-const keycloakConfig = JSON.parse(fs.readFileSync(path.resolve(__dirname, './path/to/keycloak.json')))
-const keycloakService = new KeycloakSecurityService(keycloakConfig)
-```
+The syntax for checking a realm role is `@hasRole(role: "realm:<role>")`. For example, `@hasRole(role: "realm:admin")`. Using a list of roles, it is possible to check for both client and realm roles at the same time.
+
+### Example: Using the @hasRole Directive to Apply Role Based Authorization on a Schema
 
-Create the schema using the directives provided by `KeycloakSecurityService`.
-```javascript
-const schemaDirectives = keycloakService.getSchemaDirectives()
+The following example demonstrates how the `@hasRole` directive can be used to define role based authorization on various parts of a GraphQL schema. This example schema represents publishing application like a news or blog website.
 
-const schema = makeExecutableSchema({
-  typeDefs,
-  resolvers,
-  // add directives
-  schemaDirectives
-})
 ```
+type Post {
+  id: ID!
+  title: String!
+  author: Author!
+  content: String!
+  createdAt: Int!
+}
+
+type Author {
+  id: ID!
+  name: String!
+  posts: [Post]!
+  address: String! @hasRole(role: "admin")
+  age: Int! @hasRole(role: "admin")
+}
 
-The last piece of work to do is telling Voyager server that the `KeycloakSecurityService` above will be used as the security service.
-```javascript
-const voyagerConfig = {
-  securityService: keycloakService
+type Query {
+  allPosts:[Post]!
+  getAuthor(id: ID!):Author!
 }
 
-const server = ApolloVoyagerServer(apolloConfig, voyagerConfig)
+type Mutation {
+  editPost:[Post]! @hasRole(role: ["editor", "admin"])
```

Comment:
It seems like edit will not accept any input. To make it nicer we can just rename it to `upvotePost` that may not need any input.